File: OpenGL/lwjgl-project/src/main/java/com/xpromus/tasks/delaunytriangulation/data/Circumcircle.java
package com.xpromus.tasks.delaunytriangulation.data;

public class Circumcircle {

    private final Point center;
    private final double radius;

    public Circumcircle(Point center, float radius) {
        this.center = center;
        this.radius = radius;
    }

    public Circumcircle(
        Point a,
        Point b,
        Point c
    ) {
        center = CalculateCenter(a, b, c);
        radius = CalculateRadius(center, a);
    }

    public boolean IsPointInCircumcircle(Point point) {
        var deltaX = point.getX() - center.getX();
        var deltaY = point.getY() - center.getY();
        var distance = Math.pow(deltaX, 2) + Math.pow(deltaY, 2);
        var radiusSquared = Math.pow(radius, 2);

        return distance < radiusSquared;
    }

    private double CalculateRadius(
        Point center,
        Point a
    ) {
        return Math.sqrt(
            Math.pow(center.getX() - a.getX(), 2) + Math.pow(center.getY() - a.getY(), 2)
        );
    }

    private Point CalculateCenter(
        Point a,
        Point b,
        Point c
    ) {
        var d = Math.abs(
            2 * (a.getX() * (b.getY() - c.getY()) + b.getX() *
                (c.getY() - a.getY()) + c.getX() *
                (a.getY() - b.getY()))
        );

        var uX = (
            (Math.pow(a.getX(), 2) + Math.pow(a.getY(), 2)) * (b.getY() - c.getY()) +
            (Math.pow(b.getX(), 2) + Math.pow(b.getY(), 2)) * (c.getY() - a.getY()) +
            (Math.pow(c.getX(), 2) + Math.pow(c.getY(), 2)) * (a.getY() - b.getY())
        ) / d;

        var uY = (
            (Math.pow(a.getX(), 2) + Math.pow(a.getY(), 2)) * (c.getX() - b.getX()) +
            (Math.pow(b.getX(), 2) + Math.pow(b.getY(), 2)) * (a.getX() - c.getX()) +
            (Math.pow(c.getX(), 2) + Math.pow(c.getY(), 2)) * (b.getX() - a.getX())
        ) / d;

        return new Point(uX, uY);
    }

    public Point getCenter() {
        return center;
    }

    public double getRadius() {
        return radius;
    }
}

File: Bowyer-Watson Visualizer/src/data/Circumcircle.ts
import { Vector3, type Triangle } from "three";

export class Circumcircle {
    readonly center: Vector3;
    readonly radius: number;

    constructor(
        center: Vector3,
        radius: number
    ) {
        this.center = center;
        this.radius = radius;
    }

    public isPointInCircumcircle(point: Vector3): boolean {
        const deltaX = point.x - this.center.x;
        const deltaY = point.y - this.center.y;
        const distance = Math.pow(deltaX, 2) + Math.pow(deltaY, 2);
        const radiusSquared = Math.pow(this.radius, 2);
        return distance < radiusSquared;
    }

    public static fromTriangle(triangle: Triangle): Circumcircle {
        const center = Circumcircle.calculateCenter(triangle.a, triangle.b, triangle.c);
        const radius = Circumcircle.calculateRadius(center, triangle.a);
        return new Circumcircle(
            center, radius
        );
    }

    public static fromPoints(a: Vector3, b: Vector3, c: Vector3): Circumcircle {
        const center = Circumcircle.calculateCenter(a, b, c);
        const radius = Circumcircle.calculateRadius(center, a);
        return new Circumcircle(
            center, radius
        );
    }

    private static calculateRadius(center: Vector3, a: Vector3): number {
        return Math.sqrt(
            Math.pow(center.x - a.x, 2) + Math.pow(center.y - a.y, 2)
        );
    }

    private static calculateCenter(a: Vector3, b: Vector3, c: Vector3): Vector3 {
        const d = Math.abs(
            2 * (
                ((a.x * (b.y - c.y)) + b.x) * 
                (c.y - a.y + c.x) *
                (a.y - b.y)
            )
        )

        const ux = (
            (Math.pow(a.x, 2) + Math.pow(a.y, 2)) * (b.y - c.y) +
            (Math.pow(b.x, 2) + Math.pow(b.y, 2)) * (c.y - a.y) +
            (Math.pow(c.x, 2) + Math.pow(c.y, 2)) * (a.y - b.y)
        ) / d;

        const uy = (
            (Math.pow(a.x, 2) + Math.pow(a.y, 2)) * (c.y - b.y) +
            (Math.pow(b.x, 2) + Math.pow(b.y, 2)) * (a.y - c.y) +
            (Math.pow(c.x, 2) + Math.pow(c.y, 2)) * (b.y - a.y)
        ) / d;

        return new Vector3(ux, uy, 0);
    }
}
